tests/usm/usm_api.h

```diff
@@ -0,0 +1,688 @@
+  sycl::buffer<T, 1> buf_dst;
+
+ public:
+  event_generator()
+      : rng{buf_size},
+        buf_src{arr_src.data(), rng},
+        buf_dst{arr_dst.data(), rng} {
+    arr_src.fill(T{0});
+    arr_dst.fill(T{0});
+  }
+
+  /** @brief Initialize arr_src with init_value and return sycl::event of
+   *         queue::submit()
+   *  @param value Some non-zero (non-default) value of type T
+   */
+  sycl::event init(sycl::queue &queue, T value) {
+    return queue.submit([&](sycl::handler &cgh) {
+      auto acc_src = buf_src.template get_access<sycl::access_mode::write>(cgh);
+      // single_task is used to make process long enough for testing purpose
+      // The function being tested must wait for this task to complete.
+      cgh.single_task<init_kernel_name<T, buf_size>>([=]() {
+        for (size_t i = 0; i < buf_size; ++i) {
+          acc_src[i] = value;
+        }
+      });
+    });
+  }
+
+  /** @brief Copy data from arr_src to arr_dst for future check
+   */
+  void copy_arrays(sycl::queue &queue) {
+    queue.submit([&](sycl::handler &cgh) {
+      using kernel_name = copy_arrays_kernel_name<T, buf_size>;
+      auto acc_src = buf_src.template get_access<sycl::access_mode::read>(cgh);
+      auto acc_dst = buf_dst.template get_access<sycl::access_mode::write>(cgh);
+      // Copy should be much faster then algorithm in 'init()' method to detect
+      // situation when tested function doesn't wait for passed events
+      cgh.parallel_for<kernel_name>(rng, [=](sycl::id<1> idx) {
+        const size_t i = idx[0];
+        acc_dst[i] = acc_src[i];
+      });
+    });
+  }
+
+  /** @brief Check that elements of arrays are equal to each other and to
+   *         value
+   *  @param value The same value as value passed to init() method
+   */
+  bool check(T value) {
+    bool result = true;
+    auto acc_src = buf_src.template get_access<sycl::access_mode::read>(rng);
+    auto acc_dst = buf_dst.template get_access<sycl::access_mode::read>(rng);
+    for (size_t i = buf_size; i > 0; --i) {
+      const size_t idx{i - 1};
+      result = result && (acc_src[idx] == acc_dst[idx]);
+      result = result && (acc_dst[idx] == value);
+    }
+    return result;
+  }
+};
+
+/** @brief Provides the root test logic for every method, caller, number of
+ *         events and allocation type
+ *  @tparam count Size of the allocation, in number of items
+ *  @tparam alloc Allocation to use for the pointer. In case the method we
+ *          should test requires more than a single pointer allocation, this
+ *          parameter defines allocation to use for the pointer we will use to
+ *          check the method results (target pointer for the copy operations for
+ *          example)
+ */
+template <typename T, size_t count, template <typename, size_t> class checkT,
+          typename caller, allocation alloc, size_t numEvents>
+struct test {
+  static void run(sycl_cts::util::logger &log, const std::string &typeName) {
```

Comment:
I wonder whether we could just use `typeid(T).name()` instead of passing around the type string.
Of course the name might be different according to the compiler, but perhaps it is enough to track results?
Or do we need always the same message across architectures & compilers.

Reply:
I've applied this comment in this commit: 5e37167aa8fa7f1e4cb685050c9115ae96a23e06